pyosys: rewrite using pybind11

donn opened the pull request:
_What are the reasons/motivation for this change?_

pybind11 is a reasonably more lightweight library that is subjectively a bit more intuitive to use. But more importantly, it does not require an external library (boost) to be installed on the machine, and does not require boost to be compiled for every version of Python.

_Explain how this is achieved._

- Rewrite all Python features to use the pybind11 library instead of boost::python. Unlike boost::python, pybind11 is a header-only library that is just included by Pyosys code ~~saving a lot of compile time on wheels~~ [EDIT: as it turns out not really, I used enough C++ templates to offset the time gain from not compiling boost]
- Wrapper generator entirely rewritten using a fault-tolerant c++ header-parsing library
   - Most of the "translation" moved outside of the file
- Move Python-related elements to `pyosys` directory at the root of the repo
- Create really short test set for pyosys that just exercises basic functionality

### Pyosys API Breaks
- The embedded interpreter now requires `from pyosys import libyosys as ys` like wheels
- Changes in bridging semantics for some types
  - Containers are declared as "opaque types" and are passed by reference to Python - many methods have been implemented to make them feel right at home without the overhead/ambiguity of copying to Python and then copying back after mutation
  - Only classes that handle virtual methods are Monitor and Pass, which now use "trampoline" pattern to support overriding virtual methods in Python: virtual methods no longer use (or support) `py_` prefix
- Overloaded methods no longer include type information in the function name, e.g. `Pass.call__YOSYS_NAMESPACE_RTLIL_Design__std_vector_string_` is just `Pass.call` (pybind11 handles overload resolution)
- `ys.Yosys` has been renamed `ys.Globals` to more adequately communicate its purpose

_If applicable, please suggest to reviewers how they can test the change._

```bash
pip3 install -v .
python3 ./tests/pyosys/run_tests.py
```

---

Resolves #4591 


---

TODO:

- [x] rebase on main
- [x] set operations (union, intersection, difference, etc)
- [x] equivalence operators (need constexpr to determine if containers are comparable)
- [x] global variables
- [x] study ~~and address~~ memory concerns raised by @jix

gatecat commented:
Just as a data point, some years ago nextpnr moved from boost::python to pybind11 and it was a very good decision, the number of weird Python-related build issues we had dropped significantly.

donn committed:
pyosys: rewrite using pybind11

- Rewrite all Python features to use the pybind11 library instead of boost::python.
  Unlike boost::python, pybind11 is a header-only library that is just included by Pyosys code, saving a lot of compile time on wheels.
- Factor out as much "translation" code from the generator into proper C++ files
- Fix running the embedded interpreter not supporting "from pyosys import libyosys as ys" like wheels
- Move Python-related elements to `pyosys` directory at the root of the repo
- Slight shift in bridging semantics:
  - Containers are declared as "opaque types" and are passed by reference to Python - many methods have been implemented to make them feel right at home without the overhead/ambiguity of copying to Python and then copying back after mutation
  - Monitor/Pass use "trampoline" pattern to support virual methods overridable in Python: virtual methods no longer require `py_` prefix
- Create really short test set for pyosys that just exercises basic functionality

donn committed:
pyosys: rewrite wrapper generator

[skip ci]

jix commented:
@donn During the Dev JF where you proposed this, I brought up the inherent issue with the object lifetime of `Cell`s and `Wire`s being determined by RTLIL and no way for a binding to extend this. What I didn't realize during that call is that a mechanism I added to extend the lifetime for Cells tracked by bufnorm could be generalized to provide safe Cell and Wire references for python bindings as well.

The idea would roughly be to add a `dict<RTLIL::NamedObject *, int> external_refcounts` together with some retain/release API. If a retained object gets `module->remove(..)`d, instead of calling `delete object` right away, we would clear the name and all other data (marking it as deleted) and move it to a deferred deletion area (so it doesn't show up when iterating over module items) until all external references are gone. This would require checking `external_refcounts` for each deleted object, but this is essentially free when there are no external references and relatively cheap when there are few. While what bufnorm currently does is more limited, we might be able to use the same mechanism across bufnorm and the overhauled python bindings, so in that case there would be no additional overhead at all if bufnorm was already in use, which should become more common over time.

donn committed:
pyosys: globals, set operators for opaque types

There is so much templating going on that compiling wrappers.cc now takes 1m1.668s on an Apple M4…

donn committed:
tests/pyosys: print log on failed test, fix make clean

donn committed:
misc: `WITH_PYTHON` -> `YOSYS_ENABLE_PYTHON`

For consistency.

Also trying a new thing: only rebuilding objects that use the pybind11 library. The idea is these are the only objects that include the Python/pybind headers and thus the only ones that depend on the Python ABI in any capacity, so other objects can be reused across wheel builds. This has the potential to cut down build times.

donn committed:
pyosys/hashlib: equivalence operators

donn committed:
pyosys: fix ref-only classes, implicit conversions

+ cleanup

donn commented:
@jix So I looked into it - noting it down for myself in the future but your implementation is basically this code block here:

https://github.com/YosysHQ/yosys/blob/7ebd9721656ba88f3cc25ef10882ed3e82e3c182/kernel/rtlil.cc#L2847

I think if that's not a blocker then yeah maybe I can tackle Wire/Cell lifecycles in a separate PR.

KrystalDelusion commented:
@donn how would you feel about writing some documentation for using pyosys?  If you're not familiar with the rst format that we use, markdown would also be fine and then I can convert it.  At the moment it looks like the only thing we have is the `pass.py` and `script.py` example scripts and the pyosys tests.

donn commented:
@KrystalDelusion I can try putting something under "Using Yosys (advanced)."

jix commented:
> your implementation is basically this code block here:

That's correct. Right now it just does what bufnorm needs, but ideally we would end up with a single mechanism that allows different parts of yosys to "retain" and "release" named objects within a module. That can then be shared by pyosys and bufnorm. Let me know when you want to work on this, so we can discuss the requirements for a more general way to do this and how to best implement that.

donn committed:
docs: write small guide for using pyosys

KrystalDelusion reviewed (commented):
Thanks for the documentation!  I have left a couple minor comments on some of the wording, but otherwise it looks great.

donn commented on docs/source/using_yosys/pyosys.rst:
```diff
@@ -0,0 +1,197 @@
+
+2. Installing the Pyosys wheels
+
+   On macOS and GNU/Linux (specifically, not musllinux,) you can install
```
I'll just remove the exclusion altogether, it's a bit pedantic :)

donn committed:
docs: proofread pyosys.rst

Primarily address feedback from @KrystalDelusion (thanks!)

KrystalDelusion reviewed (commented):
I haven't looked at the implementation details, but the documentation looks good 👍 Thanks @donn

KrystalDelusion commented:
@donn can `libboost-python-dev` and `boost-python3` be safely removed from the prerequisites for linux/mac respectively?

donn commented:
@KrystalDelusion Yeah.
